client/router.js

```diff
@@ -353,99 +373,9 @@ Router.prototype._buildTracker = function() {
 Router.prototype._invalidateTracker = function() {
   this.safeToRun = true;
   this._tracker.invalidate();
-};
-
-Router.prototype._getRegisterTriggersFn = function(triggers) {
-  var fn = function(triggerFns, options) {
-    options = options || {};
-
-    if (options.only && options.except) {
-      var message = "triggers does not support 'only' and 'except' at the same time.";
-      throw new Error(message);
-    }
-
-    _.each(triggerFns, function(fn) {
-      if (typeof fn !== 'function') {
-        return;
-      }
-
-      if (options.only) {
-        fn._only = {};
-        _.each(options.only, function(name) {
-          fn._only[name] = 1;
-        });
-      }
-
-      if (options.except) {
-        fn._except = {};
-        _.each(options.except, function(name) {
-          fn._except[name] = 1;
-        });
-      }
-
-      triggers.push(fn);
-    });
-  };
-
-  return fn;
-};
-
-Router.prototype._shouldCallTrigger = function(current, fn) {
-  var name = current.route.name;
-  var shouldCall;
-
-  if (typeof fn !== 'function') {
-    return false;
-  }
-
-  if (fn._only) {
-    shouldCall = !!fn._only[name];
-  } else if (fn._except) {
-    shouldCall = !fn._except[name];
-  } else {
-    shouldCall = true;
-  }
-
-  return shouldCall;
-};
-
-Router.prototype._processTriggersEnter = function(current) {
-  var self = this;
-
-  _.each(this._triggersEnter, function(fn) {
-    if (self._shouldCallTrigger(current, fn)) {
-      fn(current);
-    }
-  });
-};
-
-Router.prototype._processTriggersExit = function(ctx, next) {
-  var self = this;
-
-  _.each(self._triggersExit, function(fn) {
-    if (self._shouldCallTrigger(self._current, fn)) {
-      fn(self._current);
-    }
-  });
-
-  next();
-};
-
-Router.prototype._registerRouteTriggersExit = function(route) {
-  var self = this;
-
-  if (route._triggersExit.length > 0) {
-    // add route's exit triggers
-    self._page.exit(route.path, function(ctx, next) {
-      _.each(route._triggersExit, function(fn) {
-        if (typeof fn === 'function') {
-          fn(self._current);
-        }
-      });
-
-      next();
-    });
-  }
+  // we need to trigger the above invalidations immediately
+  // otherwise, we need to face some issues with route context swapping
+  Tracker.flush();
```

Comment:
Check how flush behave if a user change the route from a template helper. 

That's weird. But possible.